libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Network/Network/Action/CSFSalesforceAction.m

```diff
@@ -118,111 +118,127 @@ - (NSDictionary *)headersForAction {
-                    }
-                    break;
+        case 401:
+            // unauthorized (not logged in / session expired)
+            // The session ID or OAuth token used has expired or is invalid.
+            // The response body contains the message and errorCode.
+            if ([errorCode isEqualToString:@"INVALID_SESSION_ID"]) {
+                requestSessionRefresh = YES;
             }
+            break;
             
-            NSString *errorDescription = [msgObj description] ?: [NSString stringWithFormat:@"HTTP %ld for %@ %@", (long)response.statusCode, self.method, self.verb];
-            NSMutableDictionary *userInfoDict = [NSMutableDictionary dictionaryWithDictionary:@{ NSLocalizedDescriptionKey:errorDescription,
-                                                                                                 CSFNetworkErrorActionKey: self,
-                                                                                                 CSFNetworkErrorAuthenticationFailureKey: @(requestSessionRefresh) }
-                                                 ];
-            if (errorCode.length > 0) {
-                userInfoDict[NSLocalizedFailureReasonErrorKey] = errorCode;
-            }
-            responseError = [NSError errorWithDomain:CSFNetworkErrorDomain
-                                                code:response.statusCode
-                                            userInfo:userInfoDict];
-        }
-        
-        // TODO: We need to figure out how to handle the security token some other way, so it doesn't collide with other action types.
-        if (!responseError && [content isKindOfClass:[NSDictionary class]]) {
-            NSDictionary *jsonContent = (NSDictionary*)content;
-            NSString *securityToken = jsonContent[CSFActionSecurityTokenKey]; // retrieve the CSRF security token
-            if (securityToken) {
-                network.securityToken = securityToken;
+        case 403:
+            // forbidden (user isn't allowed to do the operation). The request has been refused.
+            // i.e. 'operation couldn't be completed'
+            break;
+            
+        case 404:
+            // resource was not found or deleted
+            break;
+            
+        case 408:
+            // request timeout
+            // request took too long and was aborted - max time per request is a setting, when last check, 120s
+            break;
+            
+        case 503:
+            // unavailable - too many requests in an hour
+            // max concurrent requests hit - max concurrent requests is a setting, when last confirmed it was 50 per JVM)
+            break;
+            
+        case 500:
+            // all other errors: "An error has occurred within Force.com"
+            // fall-through
+            
+        default:
+            // When the user is revoked, the response object is nil (so error code is 0)
```

Comment:
Not sure, honestly.  This code is just moved from its original location in `contentFromData:fromResponse:error:`.  I figured we could further refine this error handling, once we have a better idea of the server-side contract.  But in the meantime, it should cover whatever cases it covered before.